File: Refinement/Modules/CardChoiceTrialCombination.ts
import { possibleFactions, possibleLocations, possibleHouseCards } from "../../../Data/GameConstants.js";
import { ITrialCombination, ProbablyDistribution, AnalyzedData, CardChoiceAnalysisData } from "../../../Contracts/AnalysisContracts.js";
import { HouseCard } from "../../../Data/GameTypes.js";
import { combineCardDistributionForMap, combineCardDistributions, emptyAverageCardDistributionFactory } from "../../../Utilities/MiltaryAnalysisUtilities.js";
import { AnalyzedKeys } from "../../../Contracts/AnalysisConstants.js";

export const CardChoiceTrialCombination : ITrialCombination<CardChoiceAnalysisData> = (trials, combinedData: AnalyzedData)=>{
    if (combinedData[AnalyzedKeys.CardChoices] === undefined) combinedData[AnalyzedKeys.CardChoices] = emptyAverageCardDistributionFactory()
    
    const result = combinedData[AnalyzedKeys.CardChoices]
    
    trials.forEach((trial)=>{
        possibleFactions.forEach((faction)=>{
            const trialFactionStats = trial[AnalyzedKeys.CardChoices][faction]

            possibleLocations.forEach((location)=>{
                combineCardDistributions(
                    trialFactionStats.LocationDistributions[location], 
                    result[faction].LocationDistributions[location]
                )
            })

            trialFactionStats.NonCardStrengthDistributions.forEach((probablyDistribution, strength)=>{
                combineCardDistributionForMap(result[faction].NonCardStrengthDistributions, strength, probablyDistribution)
            })

            trialFactionStats.OppositeSideStrengthDistributions.forEach((probablyDistribution, strength)=>{
                combineCardDistributionForMap(result[faction].OppositeSideStrengthDistributions, strength, probablyDistribution)
            })
        })
    })

    return result
}

File: 1_Analysis/Modules/CardChoiceTrialCombination.py
from typing import List, Dict, Any

from Contracts.AnalysisConstants import AnalyzedKeys
from Contracts.AnalysisContracts import AnalyzedData
from Data.GameConstants import possible_factions, possible_locations
from Utilities.MiltaryAnalysisUtilities import (
    combine_card_distributions,
    combine_card_distribution_for_map,
    empty_average_card_distribution_factory,
)

CardChoiceAnalysisData = Dict[str, Any]


def card_choice_trial_combination(
    trials: List[CardChoiceAnalysisData], combined_data: AnalyzedData
) -> None:
    if AnalyzedKeys.CardChoices not in combined_data:
        combined_data[AnalyzedKeys.CardChoices] = empty_average_card_distribution_factory()

    result: Dict[str, Any] = combined_data[AnalyzedKeys.CardChoices]

    for trial in trials:
        for faction in possible_factions:
            trial_faction_stats: Dict[str, Any] = trial[AnalyzedKeys.CardChoices][faction]

            for location in possible_locations:
                combine_card_distributions(
                    trial_faction_stats['LocationDistributions'][location],
                    result[faction]['LocationDistributions'][location],
                )

            for strength, prob_dist in trial_faction_stats['NonCardStrengthDistributions'].items():
                combine_card_distribution_for_map(
                    result[faction]['NonCardStrengthDistributions'],
                    strength,
                    prob_dist,
                )

            for strength, prob_dist in trial_faction_stats['OppositeSideStrengthDistributions'].items():
                combine_card_distribution_for_map(
                    result[faction]['OppositeSideStrengthDistributions'],
                    strength,
                    prob_dist,
                )
